src/mpi/group/grouputil.c

```diff
@@ -103,25 +94,25 @@ int MPIR_Group_create_map(int size, int rank, MPIR_Session * session_ptr, MPIR_L
         /* See 5.3.2, Group Constructors.  For many group routines,
          * the standard explicitly says to return MPI_GROUP_EMPTY;
          * for others it is implied */
+        MPL_free(map);
         *new_group_ptr = MPIR_Group_empty;
         goto fn_exit;
-    }
+    } else {
+        MPIR_Group *newgrp;
+        mpi_errno = MPIR_Group_create(size, &newgrp);
+        MPIR_ERR_CHECK(mpi_errno);
 
-    MPIR_Group *newgrp;
-    mpi_errno = MPIR_Group_create(size, &newgrp);
-    MPIR_ERR_CHECK(mpi_errno);
+        newgrp->rank = rank;
+        MPIR_Group_set_session_ptr(newgrp, session_ptr);
 
-    newgrp->rank = rank;
-    MPIR_Group_set_session_ptr(newgrp, session_ptr);
+        newgrp->pmap.use_map = true;
+        newgrp->pmap.u.map = map;
 
-    for (int i = 0; i < size; i++) {
-        newgrp->lrank_to_lpid[i].lpid = map[i];
+        /* TODO: build hash to accelerate MPIR_Group_lpid_to_rank */
```

Comment:
We should add a TODO for optimization that check if a `map` is actually a `stride` and convert it to `stride`. Since all `MPIR_Group` calculations internally create `map`, we probably end up with most groups using `map`.

Reply:
I had this done already. It may be in the later PR. I'll find it and move it here.

Reply:
I just saw that :)

Reply:
Moved the commit to this PR.